### PR TITLE
Publish dev container

### LINKED ISSUE #1414: Publish developer docker containers on dockerhub
See #1309

### PR DESCRIPTION
**Pull request by seanpmorgan**
Closes #1414

Discussion Points:
1. This adds a bit of computation time to the action on commit push. I don't think this is a big deal as this doesn't slow down feedback to contributors
2. There is no addons repository included in the docker image. I believe this is the correct pattern since we want users to load their own repository within the container
3. Because of 2 there is no pre-compiled bazel cache. We could add it but then the calls to bazel in the repository would need to be pointed to that cache. Given that our build only takes a couple of minutes I'd prefer to do this in another PR if we want to.

I tested the image in VScode using a dev-container and it worked as expected.

### CODE REVIEW

File: .github/workflows/build_dev_container.sh
Comment (bhack) [inline .github/workflows/build_dev_container.sh:8]: Could this be a problem if users test the dev_container image builds process locally (i.e. PRs over dev_container recepits itself)?

**Comment by bhack** [comment]
I am looking at the compressed size of devel image TF vs TFA:

https://hub.docker.com/r/tensorflow/tensorflow/tags
https://hub.docker.com/r/tfaddons/tensorflow/tags

The TF CPU devel is 1/3 of our devel image and we have 1GB of compressed overhead over tensorflow GPU devel (uncompressed our is 9.2GB).

**Comment by bhack** [comment]
I think it could be useful to have a default small size 1GB image to lower the download overhead but I think that is still not possible to handle multiple configuration (GPU/CPU) in the dev-containers https://github.com/microsoft/vscode-remote-release/issues/2143#issuecomment-600087021.

The new devcontainer support PR is at https://github.com/microsoft/vscode-dev-containers/pull/347

**Comment by bhack** [comment]
Could you add
```
sudo update-alternatives --install /usr/bin/clang++ clang++ /usr/bin/clang++-8 100
sudo update-alternatives --install /usr/bin/clang clang /usr/bin/clang-8 100
```

**Comment by bhack** [comment]
If you could set the alias also for `clang-format` it could be useful

**Comment by seanpmorgan** [comment]
> I am looking at the compressed size of devel image TF vs TFA:
> 
> https://hub.docker.com/r/tensorflow/tensorflow/tags
> https://hub.docker.com/r/tfaddons/tensorflow/tags
> 
> The TF CPU devel is 1/3 of our devel image and we have 1GB of compressed overhead over tensorflow GPU devel (uncompressed our is 9.2GB).



> I think it could be useful to have a default small size 1GB image to lower the download overhead but I think that is still not possible to handle multiple configuration (GPU/CPU) in the dev-containers [microsoft/vscode-remote-release#2143 (comment)](https://github.com/microsoft/vscode-remote-release/issues/2143#issuecomment-600087021).

As discussed offline this is a consequence of https://github.com/tensorflow/tensorflow/issues/38352 . There is not much we can do without deviating from the custom-op recommended images.

File: .github/workflows/build_dev_container.sh
Comment (seanpmorgan) [reply .github/workflows/build_dev_container.sh:8]: Yeah which is why I moved it into `.github/workflows` it is not meant to be built locally. We could make it so it calls a script in `tools/` but then the user would need download the base image which is almost as large so I don't see a lot of benefit

**Comment by seanpmorgan** [comment]
> Could you add
> 
> ```
> sudo update-alternatives --install /usr/bin/clang++ clang++ /usr/bin/clang++-8 100
> sudo update-alternatives --install /usr/bin/clang clang /usr/bin/clang-8 100
> ```

Sure, but could you expand on why its needed. All of our documentation uses gcc and nvcc to compile. I know there was work on getting TF to compile using cuda-clang but not sure if we want to encourage that?

**Comment by bhack** [comment]
I've not checked in detail where we use what but I was thinking as a general rule to not going to ovveride to much settings properties in vsocde devcontainer about specific strings or to monitor if  we change clang version cause there we are on another repository. 
I prefer that we have the control in the image we are generating so we have constant alias.
If we don't care about clang probably we could do It only for clang-format?

**Comment by bhack** [comment]
I don't know if was clear. But I prefer to not set all these properties  e.g. `"C_Cpp.clang_format_path"` to the specific `/usr/local/bin/clang-format-9`. 
If we use the default alias it is easier to maintain cause we don't need to set all the properties.

File: CONTRIBUTING.md
Comment (bhack) [inline CONTRIBUTING.md:285]: This seems that is set by default to 1 in the image/container. 
So if I follow CPU custom_op documentation it build cuda files also without exporting this env.

Comment (bhack) [reply CONTRIBUTING.md:285]: As we are using the `custom_ops` GPU image version as our baseline it is inherited by https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/ci_build/Dockerfile.custom_op_gpu#L21

**Comment by bhack** [comment]
> As discussed offline this is a consequence of [tensorflow/tensorflow#38352](https://github.com/tensorflow/tensorflow/issues/38352) . There is not much we can do without deviating from the custom-op recommended images.

I've changed `tools/docker/build_wheel.Dockerfile` using `FROM tensorflow/tensorflow:devel-gpu as base_install` and the `bazel build --enable_runfiles build_pip_pkg` TF addons compile command seems to run fine.

**Comment by bhack** [comment]
I think that we need to update a little bit https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md after merging this. Can you open a ticket as a reminder?

**Comment by bhack** [comment]
Need we still to run `bash tools/install_so_files.sh`?

File: CONTRIBUTING.md
Comment (seanpmorgan) [reply CONTRIBUTING.md:292]: Resolved since using the cpu custom-op image as base.

File: tools/docker/dev_container.Dockerfile
Comment (bhack) [inline tools/docker/dev_container.Dockerfile:9]: With this we are duplicating the tensorflow install layer.  Do you think to remove it after https://github.com/tensorflow/tensorflow/pull/40189 or need we to duplicate it for other goals?

File: tools/install_deps/clang-format.sh
Comment (bhack) [inline tools/install_deps/clang-format.sh:20]: As you want. Finally I don't think that vscode intellisense is using the executable.

File: tools/docker/dev_container.Dockerfile
Comment (seanpmorgan) [reply tools/docker/dev_container.Dockerfile:9]: Yeah remove it after that is merged and a new container is published

File: tools/docker/dev_container.Dockerfile
Comment (bhack) [inline tools/docker/dev_container.Dockerfile:2]: Here you lost the build ARGS to prepare the switch between cpu and gpu that was in https://github.com/seanpmorgan/addons/pull/14/files#diff-76fe3c36c889ccfcf62f6e35439d030fR3-R5

File: tools/build_dev_container.sh
Comment (bhack) [inline tools/build_dev_container.sh:5]: We could duplicate this command with the build_args to publish the GPU image. See the ARGS comment below.

File: .github/workflows/release.yml
Comment (bhack) [inline .github/workflows/release.yml:114]: We could duplicate this command to push the `latest` without suffixs (the is the GPU image)

**Comment by bhack** [comment]
As a reminder if you merge https://github.com/tensorflow/addons/pull/1808/ you need to add the requirements here.

File: CONTRIBUTING.md
Comment (bhack) [inline CONTRIBUTING.md:292]: I think that we need to set this only in the GPU image

**Comment by bhack** [comment]
`custom-ops` images are not produced for RC https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/ci_build/release/ubuntu_16/custom_op/release.sh#L39

File: CONTRIBUTING.md
Comment (seanpmorgan) [reply CONTRIBUTING.md:292]: Yes, but also if it's built from source locally.

**Comment by seanpmorgan** [comment]
@bhack @tensorflow/sig-addons-maintainers so there are still a few sharp edges here, but it aligns with the recent merge of SIG-IO:
https://github.com/tensorflow/io/issues/1087

I would prefer that we merge this semi-hardcoded cpu devel build and then I'll open an issue and start work on a seperate GPU build. This got a bit stale and it'd be easier to get this merged and then upgrade it IMO. Also waiting for clarity from SIG-Build about custom-op images and their future. Does that seem okay?

File: tools/docker/dev_container.Dockerfile
Comment (bhack) [reply tools/docker/dev_container.Dockerfile:9]: Ok I think that this for cpu could be removed now

**Review by bhack — APPROVED** [review]

**Pull request merged**
